// react-stars-rating/build/elements/Star.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _templateObject = _taggedTemplateLiteral(['\n  height: 40px;\n  width: 20%;\n  display: block;\n  position: relative;\n  cursor: ', ';\n  &:after {\n    transition: all 0.4s ease-out;\n    -webkit-font-smoothing: antialiased;\n    position: absolute;\n    content: \'\u2605\';\n    color: ', ';\n    top: 0;\n    left: 0;\n    width: 100%;\n    height: 100%;\n    text-align: center;\n    font-size: 14px;\n  }\n'], ['\n  height: 40px;\n  width: 20%;\n  display: block;\n  position: relative;\n  cursor: ', ';\n  &:after {\n    transition: all 0.4s ease-out;\n    -webkit-font-smoothing: antialiased;\n    position: absolute;\n    content: \'\u2605\';\n    color: ', ';\n    top: 0;\n    left: 0;\n    width: 100%;\n    height: 100%;\n    text-align: center;\n    font-size: 14px;\n  }\n']);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _colors = require('../styles/colors');

var colors = _interopRequireWildcard(_colors);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var Label = _styledComponents2.default.label(_templateObject, function (_ref) {
  var disabled = _ref.disabled;
  return disabled ? 'normal' : 'pointer';
}, function (_ref2) {
  var checked = _ref2.checked,
      theme = _ref2.theme;
  return checked ? colors[theme] : '#444444';
});

// content: ${({checked}) => checked ? '★' : '☆'};

var Star = function (_Component) {
  _inherits(Star, _Component);

  function Star() {
    var _ref3;

    var _temp, _this, _ret;

    _classCallCheck(this, Star);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref3 = Star.__proto__ || Object.getPrototypeOf(Star)).call.apply(_ref3, [this].concat(args))), _this), _this.handleForOnStarClick = function (evt) {
      var _this$props = _this.props,
          disabled = _this$props.disabled,
          index = _this$props.index,
          onClick = _this$props.onClick;


      if (!disabled) {
        onClick(index + 1);
      }
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Star, [{
    key: 'render',


    /**
     * Render method
     * @returns {Object} JSX HTML markup
     */
    value: function render() {
      var _props = this.props,
          disabled = _props.disabled,
          index = _props.index,
          checked = _props.checked,
          theme = _props.theme;

      var styles = {
        display: 'inline-block',
        width: '14px'
      };

      return _react2.default.createElement(
        'span',
        { className: 'lm-star', style: styles, onClick: this.handleForOnStarClick },
        _react2.default.createElement(Label, { checked: checked, theme: theme, disabled: disabled })
      );
    }
  }]);

  return Star;
}(_react.Component);

Star.defaultProps = {
  disabled: false,
  onClick: function onClick() {
    return false;
  },
  checked: false
};

Star.propTypes = {
  index: _propTypes2.default.number.isRequired,
  disabled: _propTypes2.default.bool,
  onClick: _propTypes2.default.func,
  checked: _propTypes2.default.bool,
  theme: _propTypes2.default.string.isRequired
};

exports.default = Star;
